README.md
# react-native-typescript-transformer

Seamlessly use TypeScript with react-native >= 0.45

## Stop! You probably don't need this package.

If you are starting a new React Native project, you can initialize the project with this command:

    react-native init MyAwesomeProject --template typescript

This will set up the project to transpile your TypeScript files using Babel.

Otherwise, if you're using React Native 0.57+ and you are converting an existing RN app to TS, then you can follow the configuration in this gist: https://gist.github.com/DimitryDushkin/bcf5a7f5df71113c67dbe2e890008308

### Babel Caveats

Babel will not type-check your files. You'll still want to use the TypeScript compiler as a kind of linter (with the `noEmit` compiler option set to true).

Also there are four rarely-used langauge features that can't be compiled with Babel.

From [this blog post](https://blogs.msdn.microsoft.com/typescript/2018/08/27/typescript-and-babel-7/):

- namespaces
- bracket style type-assertion/cast syntax regardless of when JSX is enabled (i.e. writing `<Foo>x` won’t work even in `.ts` files if JSX support is turned on, but you can instead write `x as Foo`).
- enums that span multiple declarations (i.e. enum merging)
- legacy-style import/export syntax (i.e. `import foo = require(...)` and `export = foo`)

Don't expect this list to grow.

## I'm on RN < 0.57 or I definitely want to compile my TypeScript files using TypeScript and not Babel

### Step 1: Install

    yarn add --dev react-native-typescript-transformer typescript

### Step 2: Configure TypeScript

Make sure your tsconfig.json has these compiler options:

```json
{
  "compilerOptions": {
    "target": "es2015",
    "jsx": "react",
    "noEmit": true,
    "moduleResolution": "node",
  },
  "exclude": [
    "node_modules",
  ],
}
```

See [tsconfig.json Notes](#tsconfigjson-notes) for more advanced configuration details.

### Step 3: Configure the react native packager

#### RN >= 0.59

In your projects root, extend `metro.config.js` so it contains the `transformer.babelTransformerPath` property:

```js
module.exports = {
  transformer: {
    babelTransformerPath: require.resolve('react-native-typescript-transformer')
  }
};
```

#### RN >= 0.57, < 0.59

Add this to your `rn-cli.config.js` (make one if you don't have one already):

```js
module.exports = {
  transformer: {
    babelTransformerPath: require.resolve('react-native-typescript-transformer')
  }
}
```

#### RN < 0.57

Add this to your `rn-cli.config.js` (make one if you don't have one already):

```js
module.exports = {
  getTransformModulePath() {
    return require.resolve('react-native-typescript-transformer');
  },
  getSourceExts() {
    return ['ts', 'tsx'];
  }
}
```

If you need to run the packager directly from the command line, run the following

    react-native start --config /absolute/path/to/rn-cli.config.js


### Step 4: Write TypeScript code!

Note that the platform-specific index files (index.ios.js, index.android.js, etc)
still need to be .js files, but everything else can be TypeScript.

You probably want typings for react and react-native

    yarn add --dev @types/react @types/react-native

Note that if you run `yarn tsc` it will act as a type checker rather than a compiler. Run it with `--watch` to catch dev-time errors in all files, not just the one you're editing.

### Use tslib (Optional)

    yarn add tslib

in tsconfig.json

```patch
 {
   "compilerOptions": {
+    "importHelpers": true,
   }
 }
```

Doing this should reduce your bundle size. See [this blog post](https://blog.mariusschulz.com/2016/12/16/typescript-2-1-external-helpers-library) for more details.

### Use absolute paths (Optional)

Absolute paths needs to have support from both the TypeScript compiler and the react-native packager.

This section will show you how to work with project structures like this:

```
<rootDir>
├── src
│   ├── package.json
│   ├── App.tsx
│   ├── components
│   │   ├── Banana.tsx
│   ├── index.tsx
├── index.ios.js
├── package.json
├── tsconfig.json
```

Where you want to be able to `import Banana from 'src/components/Banana'` from any .ts(x) file, regardless of its place in the directory tree.

#### TypeScript

In `tsconfig.json`:

```patch
 {
   "compilerOptions": {
+    "baseUrl": "."
   }
 }
```

#### react-native

For react-native you need to add one or more `package.json` files. These only need to contain the `"name"` field, and should be placed into any folders in the root of your project that you want to reference with an absolute path. The `"name"` field's value should be the name of the folder.  So for me, I just added one file at `src/package.json` with the contents `{"name": "src"}`.

#### Jest (Optional)

If you use Jest as a test runner, add the following in your root package.json:

```patch
 {
   "jest" {
+     "modulePaths": ["<rootDir>"]
   }
 }
```

## tsconfig.json Notes

- If you enable synthetic default imports with the `"allowSyntheticDefaultImports"` flag, be sure to set `"module"` to something like "es2015" to allow the es6 import/export syntax to pass through the TypeScript compiler untouched. Then Babel can compile those statements while emitting the necessary extra code to make synthetic default imports work properly.

  This is neccessary until TypeScript implements suport for synthetic default imports in emitted code as well as in the type checker. See [Microsoft/TypeScript#9562](https://github.com/Microsoft/TypeScript/issues/9562).

- `"target"` can be anything supported by your project's Babel configuration.

- `"jsx"` can also be `"react-native"` or `"preserve"`, which are functionally identical in the context of a react-native-typescript-transformer project. In this case, the JSX syntax is compiled by Babel instead of TypeScript

- The source map options are not useful

- You probably want to specify some base typings with the `"lib"` option. I've had success with the following:

  ```patch
   {
     "compilerOptions": {
  +    "lib": [ "es2017" ],
     }
   }
  ```
  Including the `"dom"` lib is not recommended. The React Native JavaScript runtime does not include any DOM-related APIs. See [JavaScript Environment](https://facebook.github.io/react-native/docs/javascript-environment) for more details on what web APIs React Native supports.

## Jest notes

Follow the react-native setup guide for [ts-jest](https://github.com/kulshekhar/ts-jest).

Alternatively, if you want to use exactly the same transformation code for both Jest and react-native check out [this comment](https://github.com/ds300/react-native-typescript-transformer/issues/21#issuecomment-330148700).

Note that there have been no reports of problems arising from differences between code compiled by the `ts-jest` transformer and code compiled by `react-native-typescript-transformer`. Additionally, `ts-jest` takes care of a lot of edge cases and is more configurable.

## Avoid cyclical dependencies

If you're transitioning an app from `tsc` to `react-native-typescript-transformer`, you might see runtime errors which involve imported modules being `undefined`. You almost certainly have cyclical inter-module dependencies which manifest during your app's initialization. e.g. if ModuleA is `undefined` in ModuleB it means that ModuleA (in)directly imports ModuleB.

`tsc` seems to be able to mitigate some instances of these cyclical dependencies when used as a whole-app compiler. Unfortunately the module-at-a-time compilation approach that react-native's bundler supports does not permit the same optimizations.

Be especially careful of "umbrella export" files which can easily introduce these cycles.

## License

MIT

[![Empowered by Futurice's open source sponsorship program](https://img.shields.io/badge/sponsor-futurice-ff69b4.svg)](http://futurice.com/blog/sponsoring-free-time-open-source-activities?utm_source=github&utm_medium=spice&utm_campaign=react-native-typescript-transformer)


CHANGELOG.md
## 1.2.13

- Add support for "extends" tsconfig option. Contribution by @alechill

## 1.2.12

- Add support for RN >= 59

## 1.2.11

- David Sheldrick update lockfile
- David Sheldrick update readme for babel 7
- Koen Punt remove resolver.sourceExts
- Koen Punt update readme for latest metro bundler
- aleclarson fix: add missing dependency
- Andrew Goodale Simplify lib to just `es2017`
- Andrew Goodale Do not recommend "dom" TS library

## 1.2.10

Add support for RN >= 56

Contribution by timwangdev

## 1.2.9

Revert to old (sync, non-wasm) version of source-map. Not worth the trouble right now?

## 1.2.8

Remove engingeStrict field in package.json, since some people claim that the version range
`>=8.0.0` is being mishandled by npm (???)

## 1.2.7

Add enginges field in package.json, since WebAssembly is only available in node 8+

## 1.2.6

Update source-map, which now uses WASM and is apparently a lot faster.

## 1.2.5

Improve error message when failure to find tsconfig.json file

Contribution by @vyshkant in #58

## 1.2.4

Fix tsconfig.json resolution for Monorepos.

Contribution by @ali-hk in #54

## 1.2.3

Fix react native version sniffing.

Contribution by @wsxyeah

## 1.2.2

Remove `crypto` dependency. It's bundled with node.

## 1.2.1

Fix minor bug with ast source map transformation

## 1.2.0

Add implementation for getCacheKey

## 1.1.7

Add source map support for RN >= 0.52

## 1.1.6

Add basic support for RN >= 0.52

Contribution by @olofd

## 1.1.5

Borked publish, do not use.

## 1.1.4

Improve README.md

## 1.1.3

Add support for RN => 0.47

## 1.1.2

Make typescript a peer dependency of react-native-typescript-transformer

## 1.1.1

Amend support for react-native >= 0.46 in light of metro-bundler changes
before the final 0.46.0 was released.

Contribution by @petejkim

## 1.1.0

Add tentative support for react-native >= 0.46, which uses the external
`metro-bundler` package instead of an internal bundler

Contribution by @Igor1201

## 1.0.13

Fix readme

## 1.0.12

Add support for raw mappings

Contribution by @stackia

## <= 1.0.11

Consult git log. Sorry.

Contributions by @cliffkoh (json5) and @orta (typo fix)


__specs__/extends.test.js
const transformer = require('../')
const path = require('path')

describe('Config extension', () => {
  describe('Given a tsconfig that extends a relative config file', () => {
    describe('When tsconfig is loaded', () => {
      let config

      beforeAll(() => {
        process.env.TSCONFIG_PATH = path.resolve(
          __dirname,
          './__fixtures__/tsconfigExtendsFile.json'
        )
        config = transformer.loadTSConfig()
      })

      it('Should contain compiler options that are only defined in top level config', () => {
        expect(config.compilerOptions.tsconfigExtendsFile).toBe(true)
      })

      it('Should retain compiler options defined in base config that are not overriden', () => {
        expect(config.compilerOptions.tsconfigNoExtends).toBe(true)
      })

      it('Should override duplicate compiler options with values from highest level config', () => {
        expect(config.compilerOptions.overrideMe).toBe('tsconfigExtendsFile')
      })

      afterAll(() => {
        delete process.env.TSCONFIG_PATH
      })
    })
  })

  describe('Given a tsconfig that extends a multiple config files', () => {
    describe('When tsconfig is loaded', () => {
      let config

      beforeAll(() => {
        process.env.TSCONFIG_PATH = path.resolve(
          __dirname,
          './__fixtures__/tsconfigExtendsFileRecursive.json'
        )
        config = transformer.loadTSConfig()
      })

      it('Should contain compiler options that are only defined in top level config', () => {
        expect(config.compilerOptions.tsconfigExtendsFileRecursive).toBe(true)
      })

      it('Should retain compiler options in middle config that are not overriden', () => {
        expect(config.compilerOptions.tsconfigExtendsFile).toBe(true)
      })

      it('Should retain compiler options defined in base config that are not overriden', () => {
        expect(config.compilerOptions.tsconfigNoExtends).toBe(true)
      })

      it('Should override duplicate compiler options with values from highest level config', () => {
        expect(config.compilerOptions.overrideMe).toBe(
          'tsconfigExtendsFileRecursive'
        )
      })

      afterAll(() => {
        delete process.env.TSCONFIG_PATH
      })
    })
  })

  describe('Given a tsconfig that extends a module config', () => {
    describe('When tsconfig is loaded', () => {
      let config

      beforeAll(() => {
        process.env.TSCONFIG_PATH = path.resolve(
          __dirname,
          './__fixtures__/tsconfigExtendsModule.json'
        )
        config = transformer.loadTSConfig()
      })

      afterAll(() => {
        delete process.env.TSCONFIG_PATH
      })

      it('Should contain compiler options that are only defined in top level config', () => {
        expect(config.compilerOptions.tsconfigExtendsModule).toBe(true)
      })

      it('Should retain compiler options defined in base config that are not overriden', () => {
        expect(config.compilerOptions.strict).toBe(true)
      })

      it('Should override duplicate compiler options with values from highest level config', () => {
        expect(config.compilerOptions.moduleResolution).toBe('ES6')
      })
    })
  })

  describe('Given a tsconfig that extends multiple module configs', () => {
    describe('When tsconfig is loaded', () => {
      let config

      beforeAll(() => {
        process.env.TSCONFIG_PATH = path.resolve(
          __dirname,
          './__fixtures__/tsconfigExtendsModuleRecursive.json'
        )
        config = transformer.loadTSConfig()
      })

      afterAll(() => {
        delete process.env.TSCONFIG_PATH
      })

      it('Should contain compiler options that are only defined in top level config', () => {
        expect(config.compilerOptions.tsconfigExtendsModuleRecursive).toBe(true)
      })

      it('Should retain compiler options in middle config that are not overriden', () => {
        expect(config.compilerOptions.strict).toBe(true)
      })

      it('Should retain compiler options defined in base config that are not overriden', () => {
        expect(config.compilerOptions.jsx).toBe('preserve')
      })

      it('Should override duplicate compiler options with values from highest level config', () => {
        expect(config.compilerOptions.moduleResolution).toBe('ES6')
      })
    })
  })

  describe('Given a tsconfig that does not extend', () => {
    describe('When tsconfig is loaded', () => {
      let config

      beforeAll(() => {
        process.env.TSCONFIG_PATH = path.resolve(
          __dirname,
          './__fixtures__/tsconfigNoExtends.json'
        )
        config = transformer.loadTSConfig()
      })

      afterAll(() => {
        delete process.env.TSCONFIG_PATH
      })

      it('Should contain compiler options that are only defined in top level config', () => {
        expect(config.compilerOptions.tsconfigNoExtends).toBe(true)
      })

      it('Should override duplicate compiler options with values from highest level config', () => {
        expect(config.compilerOptions.overrideMe).toBe('tsconfigNoExtends')
      })
    })
  })
})


__specs__/transformer.test.js
const file = `
type Cheese = {
  readonly cheese: string
}

export default function Cheese(): Cheese {
  return {cheese: 'stilton'};
}
`

const transformer = require('../')

describe('the transformer', () => {
  it('works for production mode', async () => {
    const result = await transformer.transform(file, 'blah.tsx', {
      generateSourceMaps: true,
    })
    expect(result.code).toMatchSnapshot()
    expect(result.map).toMatchSnapshot()
  })

  it('works for dev mode', async () => {
    const result = await transformer.transform(file, 'blah.tsx')
    expect(result.code).toMatchSnapshot()
    expect(result.map).toMatchSnapshot()
  })

  it('throws errors for bad syntax', async () => {
    try {
      await transformer.transform(badSyntaxFile, 'badSyntax.tsx', {})
    } catch (e) {
      expect(e.message).toMatchSnapshot()
    }
  })

  it('does not throw errors for bad types', async () => {
    const result = await transformer.transform(badTypeFile, 'badType.tsx', {})
    expect(result.code).toMatchSnapshot()
  })
})

const badSyntaxFile = `
const x == 7;
`

const badTypeFile = `
const x: boolean = 5;
`


createTransformer.js
'use strict'
const ts = require('typescript')
const fs = require('fs')
const findRoot = require('find-root')
const os = require('os')
const path = require('path')
const process = require('process')
const semver = require('semver')
const traverse = require('babel-traverse')
const crypto = require('crypto')
const chalk = require('chalk')

const TSCONFIG_PATH = process.env.TSCONFIG_PATH

let upstreamTransformer = null

const reactNativeVersionString = require('react-native/package.json').version
const reactNativeMinorVersion = semver(reactNativeVersionString).minor

if (reactNativeMinorVersion >= 52) {
  upstreamTransformer = require('metro/src/transformer')
} else if (reactNativeMinorVersion >= 47) {
  upstreamTransformer = require('metro-bundler/src/transformer')
} else if (reactNativeMinorVersion === 46) {
  upstreamTransformer = require('metro-bundler/build/transformer')
} else {
  // handle RN <= 0.45
  const oldUpstreamTransformer = require('react-native/packager/transformer')
  upstreamTransformer = {
    transform({ src, filename, options }) {
      return oldUpstreamTransformer.transform(src, filename, options)
    },
  }
}

const { SourceMapConsumer, SourceMapGenerator } = require('source-map')

function loadJsonFile(jsonFilename) {
  if (!fs.existsSync(jsonFilename)) {
    throw new Error(`Input file not found: ${jsonFilename}`)
  }

  const buffer = fs.readFileSync(jsonFilename)
  try {
    let jju = require('jju')
    return jju.parse(buffer.toString())
  } catch (error) {
    throw new Error(
      `Error reading "${jsonFilename}":${os.EOL}  ${error.message}`
    )
  }
}

// only used with RN >= 52
async function sourceMapAstInPlace(tsMap, babelAst) {
  const tsConsumer = await new SourceMapConsumer(tsMap)
  traverse.default.cheap(babelAst, node => {
    if (node.loc) {
      const originalStart = tsConsumer.originalPositionFor(node.loc.start)
      if (originalStart.line) {
        node.loc.start.line = originalStart.line
        node.loc.start.column = originalStart.column
      }
      const originalEnd = tsConsumer.originalPositionFor(node.loc.end)
      if (originalEnd.line) {
        node.loc.end.line = originalEnd.line
        node.loc.end.column = originalEnd.column
      }
    }
  })
}

async function composeRawSourceMap(tsMap, babelMap) {
  const tsConsumer = await new SourceMapConsumer(tsMap)
  const composedMap = []
  babelMap.forEach(
    ([generatedLine, generatedColumn, originalLine, originalColumn, name]) => {
      if (originalLine) {
        const tsOriginal = tsConsumer.originalPositionFor({
          line: originalLine,
          column: originalColumn,
        })
        if (tsOriginal.line) {
          if (typeof name === 'string') {
            composedMap.push([
              generatedLine,
              generatedColumn,
              tsOriginal.line,
              tsOriginal.column,
              name,
            ])
          } else {
            composedMap.push([
              generatedLine,
              generatedColumn,
              tsOriginal.line,
              tsOriginal.column,
            ])
          }
        }
      }
    }
  )
  return composedMap
}

async function composeSourceMaps(
  tsMap,
  babelMap,
  tsFileName,
  tsContent,
  babelCode
) {
  const tsConsumer = await new SourceMapConsumer(tsMap)
  const babelConsumer = await new SourceMapConsumer(babelMap)
  const map = new SourceMapGenerator()
  map.setSourceContent(tsFileName, tsContent)
  babelConsumer.eachMapping(
    ({
      source,
      generatedLine,
      generatedColumn,
      originalLine,
      originalColumn,
      name,
    }) => {
      if (originalLine) {
        const original = tsConsumer.originalPositionFor({
          line: originalLine,
          column: originalColumn,
        })
        if (original.line) {
          map.addMapping({
            generated: {
              line: generatedLine,
              column: generatedColumn,
            },
            original: {
              line: original.line,
              column: original.column,
            },
            source: tsFileName,
            name: name,
          })
        }
      }
    }
  )
  return map.toJSON()
}

let tsConfig
let compilerOptions

module.exports.getCacheKey = function() {
  const upstreamCacheKey = upstreamTransformer.getCacheKey
    ? upstreamTransformer.getCacheKey()
    : ''
  var key = crypto.createHash('md5')
  key.update(upstreamCacheKey)
  key.update(fs.readFileSync(__filename))
  key.update(JSON.stringify(tsConfig))
  return key.digest('hex')
}

module.exports.createTransformer = customOptions => {
  customOptions = customOptions || {}
  tsConfig = (() => {
    const customTsConfigPath = customOptions.tsConfigPath || TSCONFIG_PATH
    if (customTsConfigPath) {
      const resolvedTsconfigPath = path.resolve(
        process.cwd(),
        customTsConfigPath
      )
      if (fs.existsSync(resolvedTsconfigPath)) {
        return loadJsonFile(resolvedTsconfigPath)
      }
      console.warn(
        'tsconfig file specified by tsConfigPath option or TSCONFIG_PATH environment variable was not found'
      )
      console.warn(`tsConfigPath: ${customOptions.tsConfigPath}`)
      console.warn(`TSCONFIG_PATH = ${TSCONFIG_PATH}`)
      console.warn(`resolved = ${resolvedTsconfigPath}`)
      console.warn('looking in app root directory')
    }

    const expectedTsConfigFileName = 'tsconfig.json'

    let root
    try {
      root = findRoot(process.cwd(), dir => {
        return fs.existsSync(path.join(dir, expectedTsConfigFileName))
      })
    } catch (error) {
      console.error(`${chalk.bold(`***ERROR***`)} in react-native-typescript-transformer
    
    ${chalk.red(`  Unable to find a "${expectedTsConfigFileName}" file.`)}
    
    It should be placed at the root of your project.
    Otherwise, you can specify another location using the TSCONFIG_PATH environment variable.

  `)
      process.exit(1)
    }

    const tsConfigPath = path.join(root, expectedTsConfigFileName)

    // the error message thrown by this is good enough on it's own
    return loadJsonFile(tsConfigPath)
  })()

  const compilerOptions = Object.assign(tsConfig.compilerOptions, {
    sourceMap: true,
    inlineSources: true,
  })

  return async function(src, filename, options) {
    if (typeof src === 'object') {
      // handle RN >= 0.46
      ;({ src, filename, options } = src)
    }

    if (filename.endsWith('.ts') || filename.endsWith('.tsx')) {
      const tsCompileResult = ts.transpileModule(src, {
        compilerOptions,
        fileName: filename,
        reportDiagnostics: true,
      })

      const errors = tsCompileResult.diagnostics.filter(
        ({ category }) => category === ts.DiagnosticCategory.Error
      )

      if (errors.length) {
        // report first error
        const error = errors[0]
        const message = ts.flattenDiagnosticMessageText(error.messageText, '\n')
        if (error.file) {
          let { line, character } = error.file.getLineAndCharacterOfPosition(
            error.start
          )
          if (error.file.fileName === 'module.ts') {
            console.error({
              error,
              filename,
              options,
            })
          }
          throw new Error(
            `${error.file.fileName} (${line + 1},${character + 1}): ${message}`
          )
        } else {
          throw new Error(message)
        }
      }

      const babelCompileResult = await Promise.resolve(
        upstreamTransformer.transform({
          src: tsCompileResult.outputText,
          filename,
          options,
        })
      )

      if (reactNativeMinorVersion >= 52) {
        await sourceMapAstInPlace(
          tsCompileResult.sourceMapText,
          babelCompileResult.ast
        )
        return babelCompileResult
      }

      const composedMap = await (Array.isArray(babelCompileResult.map)
        ? composeRawSourceMap(
            tsCompileResult.sourceMapText,
            babelCompileResult.map
          )
        : composeSourceMaps(
            tsCompileResult.sourceMapText,
            babelCompileResult.map,
            filename,
            src,
            babelCompileResult.code
          ))

      return Object.assign({}, babelCompileResult, {
        map: composedMap,
      })
    } else {
      return upstreamTransformer.transform({
        src,
        filename,
        options,
      })
    }
  }
}


index.js
'use strict'
const ts = require('typescript')
const fs = require('fs')
const findRoot = require('find-root')
const os = require('os')
const path = require('path')
const process = require('process')
const semver = require('semver')
const traverse = require('babel-traverse')
const crypto = require('crypto')
const chalk = require('chalk')
const deepmerge = require('deepmerge')

let upstreamTransformer = null

const reactNativeVersionString = require('react-native/package.json').version
const reactNativeMinorVersion = semver(reactNativeVersionString).minor

if (reactNativeMinorVersion >= 59) {
  upstreamTransformer = require('metro-react-native-babel-transformer/src/index')
} else if (reactNativeMinorVersion >= 56) {
  upstreamTransformer = require('metro/src/reactNativeTransformer')
} else if (reactNativeMinorVersion >= 52) {
  upstreamTransformer = require('metro/src/transformer')
} else if (reactNativeMinorVersion >= 47) {
  upstreamTransformer = require('metro-bundler/src/transformer')
} else if (reactNativeMinorVersion === 46) {
  upstreamTransformer = require('metro-bundler/build/transformer')
} else {
  // handle RN <= 0.45
  const oldUpstreamTransformer = require('react-native/packager/transformer')
  upstreamTransformer = {
    transform({ src, filename, options }) {
      return oldUpstreamTransformer.transform(src, filename, options)
    },
  }
}

const { SourceMapConsumer, SourceMapGenerator } = require('source-map')

function loadJsonFile(jsonFileName) {
  try {
    const buffer = fs.readFileSync(jsonFileName)
    const jju = require('jju')
    return jju.parse(buffer.toString())
  } catch (error) {
    throw new Error(
      `Error reading "${jsonFileName}":${os.EOL}  ${error.message}`
    )
  }
}

function getFileOrModulePath(location) {
  try {
    return require.resolve(location)
  } catch (e) {}
}

function isFile(location) {
  return fs.existsSync(location)
}

// loads config file supporting recursive extendsion from files or node modules
function loadConfig(location) {
  let json
  const configPath = getFileOrModulePath(location)
  if (configPath) {
    try {
      json = loadJsonFile(configPath)
    } catch (error) {
      throw new Error(
        `Error loading config ${location}:${os.EOL}  ${error.message}`
      )
    }
  } else {
    throw new Error(`Could not load config from ${location}`)
  }

  if (typeof json.extends === 'string') {
    const relativeCandidate = path.join(
      path.dirname(configPath),
      `${json.extends}${json.extends.endsWith('.json') ? '' : '.json'}`
    )
    const extendedLocation = isFile(relativeCandidate)
      ? relativeCandidate
      : json.extends
    const extendedJson = loadConfig(extendedLocation)
    json = deepmerge(extendedJson, json, { arrayMerge: (_, source) => source })
    delete json.extends
  }
  return json
}

// only used with RN >= 52
function sourceMapAstInPlace(tsMap, babelAst) {
  const tsConsumer = new SourceMapConsumer(tsMap)
  traverse.default.cheap(babelAst, node => {
    if (node.loc) {
      const originalStart = tsConsumer.originalPositionFor(node.loc.start)
      if (originalStart.line) {
        node.loc.start.line = originalStart.line
        node.loc.start.column = originalStart.column
      }
      const originalEnd = tsConsumer.originalPositionFor(node.loc.end)
      if (originalEnd.line) {
        node.loc.end.line = originalEnd.line
        node.loc.end.column = originalEnd.column
      }
    }
  })
}

function composeRawSourceMap(tsMap, babelMap) {
  const tsConsumer = new SourceMapConsumer(tsMap)
  const composedMap = []
  babelMap.forEach(
    ([generatedLine, generatedColumn, originalLine, originalColumn, name]) => {
      if (originalLine) {
        const tsOriginal = tsConsumer.originalPositionFor({
          line: originalLine,
          column: originalColumn,
        })
        if (tsOriginal.line) {
          if (typeof name === 'string') {
            composedMap.push([
              generatedLine,
              generatedColumn,
              tsOriginal.line,
              tsOriginal.column,
              name,
            ])
          } else {
            composedMap.push([
              generatedLine,
              generatedColumn,
              tsOriginal.line,
              tsOriginal.column,
            ])
          }
        }
      }
    }
  )
  return composedMap
}

function composeSourceMaps(tsMap, babelMap, tsFileName, tsContent, babelCode) {
  const tsConsumer = new SourceMapConsumer(tsMap)
  const babelConsumer = new SourceMapConsumer(babelMap)
  const map = new SourceMapGenerator()
  map.setSourceContent(tsFileName, tsContent)
  babelConsumer.eachMapping(
    ({
      source,
      generatedLine,
      generatedColumn,
      originalLine,
      originalColumn,
      name,
    }) => {
      if (originalLine) {
        const original = tsConsumer.originalPositionFor({
          line: originalLine,
          column: originalColumn,
        })
        if (original.line) {
          map.addMapping({
            generated: {
              line: generatedLine,
              column: generatedColumn,
            },
            original: {
              line: original.line,
              column: original.column,
            },
            source: tsFileName,
            name: name,
          })
        }
      }
    }
  )
  return map.toJSON()
}

function loadTSConfig() {
  const TSCONFIG_PATH = process.env.TSCONFIG_PATH

  if (TSCONFIG_PATH) {
    const resolvedTsconfigPath = path.resolve(process.cwd(), TSCONFIG_PATH)
    if (isFile(resolvedTsconfigPath)) {
      return loadConfig(resolvedTsconfigPath)
    }
    console.warn(
      'tsconfig file specified by TSCONFIG_PATH environment variable was not found'
    )
    console.warn(`TSCONFIG_PATH = ${TSCONFIG_PATH}`)
    console.warn(`resolved = ${resolvedTsconfigPath}`)
    console.warn('looking in app root directory')
  }

  const expectedTsConfigFileName = 'tsconfig.json'

  let root
  try {
    root = findRoot(process.cwd(), dir => {
      return isFile(path.join(dir, expectedTsConfigFileName))
    })
  } catch (error) {
    console.error(`${chalk.bold(`***ERROR***`)} in react-native-typescript-transformer
  
  ${chalk.red(`  Unable to find a "${expectedTsConfigFileName}" file.`)}
  
  It should be placed at the root of your project.
        Otherwise, you can specify another location using the TSCONFIG_PATH environment variable.

`)
    process.exit(1)
  }

  const tsConfigPath = path.join(root, expectedTsConfigFileName)

  // the error message thrown by this is good enough on it's own
  return loadConfig(tsConfigPath)
}

const tsConfig = loadTSConfig()

const compilerOptions = Object.assign(tsConfig.compilerOptions, {
  sourceMap: true,
  inlineSources: true,
})

function getCacheKey() {
  const upstreamCacheKey = upstreamTransformer.getCacheKey
    ? upstreamTransformer.getCacheKey()
    : ''
  var key = crypto.createHash('md5')
  key.update(upstreamCacheKey)
  key.update(fs.readFileSync(__filename))
  key.update(JSON.stringify(tsConfig))
  return key.digest('hex')
}

function transform(src, filename, options) {
  if (typeof src === 'object') {
    // handle RN >= 0.46
    ;({ src, filename, options } = src)
  }

  if (filename.endsWith('.ts') || filename.endsWith('.tsx')) {
    const tsCompileResult = ts.transpileModule(src, {
      compilerOptions,
      fileName: filename,
      reportDiagnostics: true,
    })

    const errors = tsCompileResult.diagnostics.filter(
      ({ category }) => category === ts.DiagnosticCategory.Error
    )

    if (errors.length) {
      // report first error
      const error = errors[0]
      const message = ts.flattenDiagnosticMessageText(error.messageText, '\n')
      if (error.file) {
        let { line, character } = error.file.getLineAndCharacterOfPosition(
          error.start
        )
        if (error.file.fileName === 'module.ts') {
          console.error({
            error,
            filename,
            options,
          })
        }
        throw new Error(
          `${error.file.fileName} (${line + 1},${character + 1}): ${message}`
        )
      } else {
        throw new Error(message)
      }
    }

    const babelCompileResult = upstreamTransformer.transform({
      src: tsCompileResult.outputText,
      filename,
      options,
    })

    if (reactNativeMinorVersion >= 52) {
      sourceMapAstInPlace(tsCompileResult.sourceMapText, babelCompileResult.ast)
      return babelCompileResult
    }

    const composedMap = Array.isArray(babelCompileResult.map)
      ? composeRawSourceMap(
          tsCompileResult.sourceMapText,
          babelCompileResult.map
        )
      : composeSourceMaps(
          tsCompileResult.sourceMapText,
          babelCompileResult.map,
          filename,
          src,
          babelCompileResult.code
        )

    return Object.assign({}, babelCompileResult, {
      map: composedMap,
    })
  } else {
    return upstreamTransformer.transform({
      src,
      filename,
      options,
    })
  }
}

module.exports = {
  getCacheKey,
  loadTSConfig,
  transform,
}
